include/envoy/thread/thread.h

```diff
@@ -2,40 +2,149 @@
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
 public:
-  OptionalLockGuard(T* lock) : lock_(lock) {
-    if (lock) {
-      lock->lock();
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_) {
+      lock_->lock();
     }
   }
 
-  ~OptionalLockGuard() {
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
     if (lock_) {
       lock_->unlock();
     }
   }
 
 private:
-  T* lock_;
+  BasicLockable* lock_;
+};
+
+/**
+ * A lock guard that deals with a lock that is not locked on construction, although
+ * it is unlocked on destruction, if necessary.
+ */
+class DeferredLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is not locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  DeferredLockGuard(BasicLockable& lock) : lock_(lock), is_locked_(false) {}
+
+  /**
+   * Destruction of the DeferredLockGuard unlocks the lock, if it was locked.
+   */
+  ~DeferredLockGuard() {
+    if (is_locked_) {
+      lock_.unlock();
+    }
+  }
+
+  // Attempts to lock the mutex, if present. Returns false if no lock was taken.
+  bool tryLock() {
+    is_locked_ = lock_.tryLock();
+    return is_locked_;
+  }
+
+private:
+  BasicLockable& lock_;
+  bool is_locked_;
+};
+
+/**
+ * Implements a LockGuard that is identical to absl::ReleasableMutexLock, but takes a
+ * BasicLockable& to allow usages to be agnostic to cross-process mutexes vs. single-process
+ * mutexes.
+ */
+class SCOPED_LOCKABLE LockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  explicit LockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    lock_.lock();
+  }
+
+  /**
+   * Destruction of the LockGuard unlocks the lock.
+   */
+  ~LockGuard() UNLOCK_FUNCTION() { lock_.unlock(); }
+
+private:
+  BasicLockable& lock_; // Set to nullptr on unlock, to prevent double-unlocking.
+};
+
+/**
+ * Implements a LockGuard that is identical to absl::ReleasableMutexLock, but takes a
+ * BasicLockable& to allow usages to be agnostic to cross-process mutexes vs. single-process
+ * mutexes.
+ */
+class SCOPED_LOCKABLE ReleasableLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  explicit ReleasableLockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(&lock) {
+    lock_->lock();
+  }
+
+  /**
+   * Destruction of the LockGuard unlocks the lock, if it has not already been explicitly released.
+   */
+  ~ReleasableLockGuard() UNLOCK_FUNCTION() {
+    if (lock_ != nullptr) {
```

Comment:
Call `release()`?

Reply:
Indeed.  I had it like that, and was getting errors 'lock not held on all paths', which I thought was due to the call to `release()`.  But that turned out to be due to the way I was passing the mutex into the LOCK macros, and after I fixed that, I forgot to put this back.  Good catch.